Remove parser conflicts

jjcnn opened the pull request:
We had 8 shift/reduce conflict in the parser related to map value types.

The problem occurs in the following situation:
```
field f : Map Uint128 List Uint32 = ...
```
Without parentheses to group the value type of the map this looks like a map that takes 3 type arguments, and indeed that is how menhir's "arbitrary resolution" resolved it.

I have now changed the grammar so that it's still parsed the same way, but without conflicts.

anton-trunov commented:
Those warnings have been bothering me forever!

It seems that you need to run

```shell
make parser-messages
```

to fully update the parser messages file.

jjcnn commented:
@anton-trunov : Updated ParserFaults.messages, and added testcases for the new parser states.

I discovered a couple of unhelpful error messages along the way, which I have attempted to fix. There are still plenty of messages that could be improved, but I'll leave that for another day when I need a break from whatever it is I'm doing.

`make parser-messages` is really useful, BTW.

anton-trunov reviewed (approved):
Awesome!
